// react-beforeunload/lib/index.esm.js
import PropTypes from 'prop-types';
import { useEffect } from 'react';
import useLatest from 'use-latest';

var useBeforeunload = function useBeforeunload(handler) {
  if (handler === void 0) {
    handler = function handler() {};
  }

  if (process.env.NODE_ENV !== 'production' && typeof handler !== 'function') {
    throw new TypeError("Expected \"handler\" to be a function, not " + typeof handler + ".");
  }

  var handlerRef = useLatest(handler);
  useEffect(function () {
    var handleBeforeunload = function handleBeforeunload(event) {
      var returnValue = handlerRef.current(event); // Chrome requires `returnValue` to be set.

      if (event.defaultPrevented) {
        event.returnValue = '';
      }

      if (typeof returnValue === 'string') {
        event.returnValue = returnValue;
        return returnValue;
      }
    };

    window.addEventListener('beforeunload', handleBeforeunload);
    return function () {
      window.removeEventListener('beforeunload', handleBeforeunload);
    };
  }, []);
};

var Beforeunload = function Beforeunload(props) {
  useBeforeunload(props.onBeforeunload);
  return props.children;
};

Beforeunload.defaultProps = {
  children: null
};

if (process.env.NODE_ENV !== 'production') {
  Beforeunload.propTypes = {
    children: PropTypes.any,
    onBeforeunload: PropTypes.func.isRequired
  };
}

export { Beforeunload, useBeforeunload };
